[Feature] basic domain

boradol opened the pull request:
### PR Type
<!-- Please check the one that applies to this PR using "x". -->

- [ ] 버그수정[Bugfix]
- [x] 기능개발[Feature]
- [ ] 코드 스타일 변경[Code style update (formatting, local variables)]
- [ ] 리팩토링 [no functional changes, no api changes]
- [ ] 빌드 관련 변경
- [x] 문서 내용 변경
- [ ] Other… Please describe:

---
### 요약(Summary)
#### 문서내용 : 
- 브랜치전략 : [**Git-flow**](https://techblog.woowahan.com/2553/)를 따릅니다.

#### 기능 개발 내용 : 
- 도메인 초기 틀을 잡았습니다.
  - [패키지 구조](https://cheese10yun.github.io/spring-guide-directory/) <- 이 링크의 글을 참고하여 구성해보았습니다. 
  - User, Shop, Theme, Price, Reservation
  
- 예약 관련된 기능을 개발하고 테스트 코드를 만들었습니다.
  - UserReservation : compositeKey
  - 테스트코드 : Repository test -> UserReservation 추가되는 내용.

---
### 상세 내용(Describe your changes)
####  구현된 ERD 사진 첨부
<img width="1146" alt="image" src="https://github.com/f-lab-edu/escape/assets/58449849/836003a5-4265-4e1e-9e76-2a85852144da">



#### User
- 나중에 로그인 기능 추가하기 전 도메인 간단하게 구현.

#### Shop
- Embedded type 사용 : Address

#### Theme
- Shop과 단방향 매핑
- Price와 양방향 매핑

#### Reservation
- UserReservation 은 compositeKey사용하여 구현 : IdClass사용

#### Test
- DataJpaTest 로 예약 추가 test

---
### Issue Number or Link

---
### PR Checklist
PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.  [Commit message convention](https://velog.io/@archivvonjang/Git-Commit-Message-Convention) 참고  (Ctrl + 클릭하세요.) 
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

boradol committed:
chore(Jenkinsfile): Add Jenkinsfile

boradol committed:
chore(.jenkinsfile): Rename jenkins

boradol committed:
chore(.jenkinsfile): Delete jenkinsfile

boradol committed:
chore(.jenkinsfile): Add pipeline script jenkinsfile

boradol committed:
chore(.gitmodules): Add gitmoduels

boradol committed:
chore(jenkinsfile): Add deploy stage

boradol committed:
chore(jenkinsfile): Add ssh connect

boradol committed:
chore(jenkinsfile): Delete ssh connect

boradol committed:
chore(deploy): Add deploy-script.sh

boradol committed:
chore(jenkinsfile): Fix jenkinsfile

boradol committed:
Create PULL_REQUEST_TEMPLATE.md

boradol committed:
Merge branch 'main' into dev

boradol committed:
docs(README): Add branch strategy

boradol committed:
feat(BaseTimeEntity): Add BaseTimeEntity

boradol committed:
feat(User): Add Domain User

boradol committed:
feat(Shop): Add Domain Shop

1. Embbed Address
2. LocationType

boradol committed:
feat(Theme): Add Domain Theme

1. Price (Bidirectional Mapping)
2. ManyToOne Shop

boradol committed:
feat(Reservation): Add Domain Reservation

1. Theme (Uni-directional mapping)
2. UserReservation (Composite key)

boradol committed:
test(UserReservation): Test saveUserReservation

boradol commented on src/main/java/com/project/escape/domain/user/SocialType.java:
```diff
@@ -0,0 +1,5 @@
+package com.project.escape.domain.user;
+
+public enum SocialType {
```
MemberJoinType?

boradol replied:
- SocialProvider
- AuthProvider 중에 고민중에 있습니다.

로그인 구현 후 적용하도록 하곘습니다.

boradol commented on src/main/java/com/project/escape/domain/reservation/UserReservation.java:
```diff
@@ -0,0 +1,35 @@
+package com.project.escape.domain.reservation;
```
패키지명 reservation.user

boradol commented on src/main/java/com/project/escape/domain/shop/LocationType.java:
```diff
@@ -0,0 +1,7 @@
+package com.project.escape.domain.shop;
+
+public enum LocationType {
```
Enum에 type을 제외해도 상관 없다.

boradol commented on src/main/java/com/project/escape/domain/theme/Price.java:
```diff
@@ -0,0 +1,21 @@
+
+@Getter
+@Entity
+public class Price {
```
UnitPrice

boradol commented on README.md:
```diff
@@ -13,6 +13,10 @@
 
+---
+## 브랜치 전략
+[Git-Flow](https://techblog.woowahan.com/2553/)를 따른다
```
브랜치 전략은 `git-flow`로 정하였습니다.

boradol committed:
refactor: Remove package wildcard

boradol committed:
refactor: Apply review

1. Move package reservation.user (UserReservation)
2. Rename Class name

boradol committed:
feat(Swagger): Add swaggerConfig

host:port/swagger-ui/index.html

boradol committed:
feat(properties): Add application-local

boradol committed:
refactor(BaseTimeEntity): Add PrePersist, PreUpdate

boradol committed:
feat(GeneralResponse): CommonResponse

+ GeneralDataResponse
+ GeneralResponseCode

boradol committed:
feat(GlobalException): Add GeneralException

Add GlobalExceptionHandler
Add ExceptionResponse
Add GeneralException
+ badreqeust, notfound, internalerro

boradol committed:
feat(ResponseFactory): Add ResponseFactory

Static Function

Success Response
Exception Response

boradol committed:
feat(UserController): Fix Response

boradol committed:
feat(UserReservation): Separate Enum

boradol committed:
feat(Theme): Add Response

boradol committed:
refactor(UserReservation): Fix composite relation

boradol commented on src/main/java/com/project/escape/domain/reservation/user/UserReservationId.java:
```diff
@@ -0,0 +1,15 @@
+@NoArgsConstructor
+@AllArgsConstructor
+public class UserReservationId implements Serializable {
+    private Long user;
+    private Long reservation;
```
`UserReservation` Entity에 있는 
`user`와 `reservation`의 변수명이 같아야 매핑이 가능했습니다. 
그래서
이곳의 변수명도 `user`, `reservation` 하였습니다.

근데 사실 얘네들이 가르키고 있는건 Id 인데요.
```suggestion
    private Long userId;
    private Long reservationId;
```
이렇게 선언하면 UserReservation에 create힐 때에 SQL에러가 났습니다.

`SQLState: 07001` 이 에러였고
SQL 쿼리에 인자의 수가 맞지 않는 경우에 나는 에러였습니다.
보니까 중복키가 선언되어 에러가 났습니다.

이런 경우에 변수를 어떻게 선택해야할까요?

f-lab-kyle replied:
제가 이해한게 맞는지 모르겠지만, UserReservation 테이블을 보면 복합키로 구성되어 있던데요~
여기서 꼭 복합키로 구성되어야 할까? 라는 고민이 들긴 합니다. reservationId가 유니크 하다면 user_id는 pk가 아니라 fk가 되어도 좋을것 같아서요~ 요거 한번 고민해봐주시고 복합키가 아닌 reservationId를 uk + pk 로 가져가면 좀 쉽게 해결이 되지 않을까 생각이 되네요~

boradol replied:
아 그렇네요. 원래 reservation을 `ManyToOne`으로 설정했다가 OneToOne으로 변경하였는데요. 그냥 fk를 걸면 간단해지네요.

boradol replied:
해결한 커밋 : https://github.com/f-lab-edu/escape/pull/2/commits/638842264ca6574a1e34dcbe07f243a2a0167583

boradol commented on src/main/java/com/project/escape/global/common/BaseTimeEntity.java:
```diff
@@ -0,0 +1,34 @@
+    @LastModifiedDate
+    private LocalDateTime modifiedAt;
+
+    @PrePersist
+    public void onPrePersist() {
+        this.createdAt = LocalDateTime.now();
+        this.modifiedAt = LocalDateTime.now();
+    }
+
+    @PreUpdate
+    public void onPreUpdate() {
+        this.modifiedAt = LocalDateTime.now();
+    }
```
UserReservation 에서 create할 때, 해당 변수가 null이 되어 처리하였습니다.
많은 데이터를 처리하지 않고, 이렇게 날짜를 넣는것은 성능상 문제가 없다고 하여 적용하였는데요.
멘토님 생각에는 이러한 처리를 어떻게 생각하시나요?

f-lab-kyle replied:
이렇게 해도 좋고 그냥 해당 필드에 기본값으로 처리하면 안되나 싶네요 ㅎ
그리고 LocalDateTime.now() 를 각각 사용하다보면 원하지 않는 사태(?)가 발생할수 있으니 주의가 필요해 보이고요
> 생성하는 시점이 23:59:59 인데 타이밍 이슈로 modifiedAt 은 00:00:00으로 들어가 비즈니스 로직에 문제가 생길수도?

boradol replied:
`@EnableJpaAuditing`을 Application 최상단에 붙여주었더니 null문제가 해결되었습니다.


[참고링크](https://hudi.blog/spring-data-jpa-auditing-create-update-date/)
```
 @EntityListeners(AuditingEntityListener.class) 이는 Auditing을 사용하겠다고 선언을 한 것인데, 
이 한줄 추가한것만으로는 활성화를 할 수 없다. 
이를 활성화하기 위해서는 Enable을 해주어야 하는데, 
이를 Application.java에 EnableJpaAuditing 어노테이션을 추가해주면 된다.
```

boradol replied:
해결한 커밋 : https://github.com/f-lab-edu/escape/pull/2/commits/b2f8a4c9735faa02ecea0ff233ebb1f880a4ec98

boradol commented on src/main/resources/application-local.yml:
```diff
@@ -0,0 +1,12 @@
+spring:
+  datasource:
+    driver-class-name: com.mysql.cj.jdbc.Driver
+    url: jdbc:mysql://localhost:3306/test?useSSL=false&allowPublicKeyRetrieval=true&serverTimezone=Asia/Seoul&characterEncoding=utf8
+    username: root
+    password: root1234
+  jpa:
+    properties:
+      hibernate:
+        ddl-auto: update
+        show_sql: true
+    generate-ddl: true
```
profile이 local일때 로컬PC에서 DB연결하려고 추가한 내용입니다.
서버정보는 들어가지 않았습니다.

boradol commented on src/main/java/com/project/escape/global/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,28 @@
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+
+@RestControllerAdvice
+public class GlobalExceptionHandler extends ResponseEntityExceptionHandler {
```
일단 StatusCode에 따라 핸들러를 만들었는데요.
Exception은 어디까지 만들어야 하며 핸들러는 주로 어떻게 나눠서 조직하는지 궁금합니다.

f-lab-kyle replied:
탑다운으로 생각하시면 좋을것 같아요.
일단 공통 예외로 처리하되 도메인을 개발하다보면 약간의 묶음 형태로 구분되는 경우가 생기는데요.
그때(비슷한것들이 생겨날때) depth를 나누어서 처리해도 무방할것 같습니다.

boradol replied:
탑다운 방식에 대해서는 멘토링 시간에 더 질문드리도록 하겠습니다. 
사실 이런식으로 프로젝트때 진행해보지 않아서 
depth를 나눈다는 의미에 대해 궁금합니다.

boradol committed:
feat(Reservation): Add soft delete

boradol commented on src/main/java/com/project/escape/domain/reservation/Reservation.java:
```diff
@@ -0,0 +1,38 @@
+    @Enumerated(EnumType.STRING)
+    private ReservationStatus status = ReservationStatus.WAITING;
+    @Column(nullable = false)
+    private Boolean isDeleted = false;
```
데이터 row를 삭제하는 것보다 (hard delete)
이렇게 flag를 둬서 하는 삭제(soft delete)방법을 선택하였습니다.
그래서 조회시 마다 삭제되지 않은 (false)인 예약 리스트를 보여주려고 하였는데요.

일단 soft delete를 선택한 이유는
1. 아무래도 reservation이랑 묶여있는 연관관계들에 user까지 깊게 연관되어 있다보니 
cascade를 하였을 때 삭제에 대한 사이드 이펙트가 일어나는 것이 염려되었습니다.
2. 그리고 soft delete를 이용하여 실제데이터가 실수로 삭제되는 것에 대해서 복구하기 쉬운 점도 있었습니다.
3. 연관관계가 복잡한 데이터를 delete쿼리를 사용해봤고 실제 사이드 이펙트를 경험하였었습니다. 그 후로 cascade에 대해 공부해보았지만, 잘못된 설정으로 한번 사라진 데이터를 복구하는데 많은 시간을 소모하였던 경험이 있어 이렇게 적용하게 되었습니다.

softdelete 처리에 대해 어떻게 생각하시나요?

f-lab-kyle replied:
softdelete에 관한 최근에 봤던 글이 있어 먼저 공유 드리고요 ㅎ https://careerly.co.kr/comments/101304

지금처럼 그렇게 하셔도 좋은데 만약 "복구"가 중요하다면 저라면 hard delete를 하면서 별도의 히스토리 테이블에 삭제한 이력을 쌓을것 같습니다.
(요건 개인 취향일수 있는데요) soft delete를 하게 되면 데이터가 점점 늘어날때 필터처리를 신경써야 하는 (이게 곧 성능에 영향을 줄수 있는) 문제가 생길것 같아 조금 심플하게 설계하는 편이긴 해요 ㅎ

boradol replied:
저도 이 글을 본 후에 softdelete에 대해서 생각을 해보게 된 것 같습니다.
근데 제 설계에서 softdelete상황이 과연 많은 문제가 될것인가에 대해 고민이 들게 되었던것 같습니다.
일단 리뷰 주신 내용으로 좀더 고민해보고 수정해보겠습니다 감사합니다.

boradol committed:
feat(UserReservation): Add CreateUserReservation & getList, getOne

1. Add Response

boradol commented on src/main/java/com/project/escape/domain/reservation/ReservationController.java:
```diff
@@ -0,0 +1,42 @@
+    private final ReservationService reservationService;
+
+    @PostMapping("/user")
+    @ResponseStatus(HttpStatus.CREATED)
+    public GeneralDataResponse<UserReservationResponse> createUserReservation(@RequestBody UserReservationRequest request) {
+        return created(reservationService.createUserReservation(request));
```
created 될 때 ResponseStatus를 Controller말고도 한곳에서 효과적으로 설정하는 밥법이 있을까요?

f-lab-kyle replied:
응답 코드를 한곳에서 공통 처리 하고싶은 상황일까요?
그렇다면 인터셉터나 필터를 고민해봐도 좋을것 같은데 장/단점이 있을것 같긴 하네요.

boradol replied:
네네 Controller에서 어노테이션 선언 말고도 공통으로 처리하고 싶은거였는데요.
인터셉터나 필터를 고려해서 구현해야 하는것이라면 
앞으로 추가될 것도 있을것 같은 필터에 만드는 것보다 그냥 그떄그때 선언하는것이 훨씬 나을것 같습니다.

boradol commented on src/main/java/com/project/escape/domain/reservation/ReservationService.java:
```diff
@@ -0,0 +1,72 @@
+        return new UserReservation(user, reservation);
+    }
+
+    private Reservation getReservation(Long reservationId) {
+        return reservationRepository.findById(reservationId)
+                .orElseThrow(ResponseFactory::notFound);
+    }
+
+    private User getUser(Long userId) {
+        return userRepository.findById(userId).orElseThrow(ResponseFactory::notFound);
+    }
```
서비스계층에서 다른 도메인은 서비스계층의 의존성 주입을 해야하는것은 알고 있는데요.
예를들어
User서비스에서는 UserResponse Dto로 반환하게되면

이 서비스에서 UserResponse -> User Entity 로 변환하는 과정들을 겪게 됩니다.
그럼 이 부분에 Entity로 효과적으로 반환하는 방법이 있을까요?

f-lab-kyle replied:
UserResponse 내부에 toEntity 같은 내부 메소드를 만들어도 좋을것 같네요. 어찌되었든 DTO라는 객체를 관련 entity로 변환하는 책임은 DTO 에게 있으니까요.

boradol replied:
해결한 커밋 : https://github.com/f-lab-edu/escape/pull/2/commits/8b743eeae8f5d48d1e5c63ccac74ad778cc3a107

boradol commented on src/main/java/com/project/escape/domain/reservation/ReservationService.java:
```diff
@@ -0,0 +1,72 @@
+    @Transactional
+    public UserReservationResponse createUserReservation(UserReservationRequest request) {
+        if (isExistsByReservationId(request)) {
+            throw badRequest(GeneralResponseCode.RESERVED_THEME_ERROR);
```
exception에 관련된것을 ResponseFactory로 만들었습니다.
에러처리를 간편하게 하기 위해 이렇게 구성하였는데요.
근데 뭔가 불편한것이 서비스계층에 badRequest라는 것이 이 해당 계층에 있어도 되는것인지 궁금합니다.

f-lab-kyle replied:
이상한 부분이 없어 보이는데요~ 말씀하신 것처럼 예외를 만드는 팩토리 클래스로 책임을 모았고
그걸 static 하게 사용하니 괜찮아 보입니다. (훨씬 더 가독성에 좋아보이고요)

boradol commented on src/test/java/com/project/escape/domain/reservation/UserReservationRepositoryTest.java:
```diff
@@ -0,0 +1,52 @@
+@ActiveProfiles(value = "local")
+@DataJpaTest
+@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
+class UserReservationRepositoryTest {
```
테스트코드는 좀더 보완해서 올리도록 하겠습니다.ㅠㅠ

boradol reviewed (commented):
안녕하세요 멘토님.

코드 작성함에 있어 사소한 것인데도 왜이렇게 고민이 많은지 모르겠습니다.ㅠㅠ
코멘트 남겼고, 혹시 그 이외에도 크게 문제되는 점 있으면 말씀 부탁드립니다.

일단, 지난번 멘토링때 패키지 만으로도 많은 이야기를 나눌 수 있어 좋았습니다.
제 안의 근거가 많이 부족하여 실제 서비스때 정하지 못한 부분이 있었는데 
그래도 멘토님과 이야기하며 확립된것 같아 좋습니다.

이번 공부를 하면서 코드 하나에 고민이 많은 것 같습니다.
너무 모르고 구현하기에 급급한 저의 지난 세월이었던 것 같습니다.

지난 멘토링 이후 저는 기본을 잘 잡아야 하는 사람같습니다.
그래서 코드량도 많지않고, 정말 기초적인 수준의 코드이지만 여기서도 많이 배울것이 있다 생각하여 
궁금한 부분에 코멘트 남깁니다.

테스트 코드를 작성하려고 하였으나.. 
1. 뭔가 잘 짜서 올리고싶다.
2. 아직 테스트할만한 정도의 재미있는 로직이 없는것 같다.
라고 생각되어 기능이 조금더 확장되어지면 그때 추가해보도록 하겠습니다.

멘토님 많이 부족한 실력입니다. 리뷰 잘 부탁드립니다.🙏🏻

f-lab-kyle commented:
그리고 하나의 PR 에서 계속 변경이 발생하면 리뷰하는 타이밍이 애매해지니
차라리 큰 피쳐 브랜치를 만들고 거기에 작게작게 PR merge를 하는 형태가 더 좋지 않을까 생각 됩니다~

boradol commented:
> 그리고 하나의 PR 에서 계속 변경이 발생하면 리뷰하는 타이밍이 애매해지니 차라리 큰 피쳐 브랜치를 만들고 거기에 작게작게 PR merge를 하는 형태가 더 좋지 않을까 생각 됩니다~

네네 맞습니다. 이런거 하나하나도 배우고 싶습니다.
일단, 멘토님이 리뷰주신 것 수정 간략하게 하고 이번 PR은 머지하도록 하겠습니다.
다음 PR부터 그렇게 적용하곘습니다.

감사합니다.

boradol committed:
refactor(UserReservation): Fix userId primary key -> foreign key

boradol committed:
refactor(BaseTimeEntity): Add EnableJpaAuditing

boradol committed:
refactor(User): Fix User toEntity

boradol committed:
refactor : Fix data rename

1. Table name : Shop -> Store
2. Embedded type : Theme.themeinfo -> difficulty, fear, activity
3. field rename : Theme.duration -> Theme.playtime
ThemeInfo.horror -> ThemeInfo.fear

boradol committed:
feat(UserReservation): Add validate logic

1. Add playerCount
2. validate PlayerCount logic
3. calculate Price logic
4. Add price in price
